### ISSUE TITLE
KYC flow in the TF app

### ISSUE THREAD

**Issue by dvorkbod** [threefoldfoundation/app_backend#197]
KYC/AML checks were at first planned to be done by and the results stored in IYO. See https://github.com/itsyouonline/identityserver/issues/661
However, due to a growing understanding of KYC/AML regulations (e.g., they still need human interpretation and judgment), it seems more appropriate to task TF and the TF app to be the ones managing this responsibility. Let's describe how.
> Comments are in text boxes, like this one. They are mostly "nice-to-haves", "long-term implications" and/or "potential issues, if the stars align".

In order to make KYC/AML a TF responsibility, we need to construct an KYC flow in the TF app. Users who already signed contracts will be asked to go through this flow, new users will be redirected to this flow after having indicated how many tokens they want to buy, before they get a contract to sign.
>Alternatively, we have people sign a contract that has been updated to state that our confirmation is pending KYC/AML approval of their info. However, this potentially makes us have to be able to defend our decision, lest we are accused of using KYC/AML failures as an easy way to get out from under contracts. Let's not try to solve this before we can or make an issue where there is none.

## Buying tokens in the age of KYC
A user wants to buy tokens. He indicates how many and how he wants to pay. Before sending a contract, we verify if this user has already passed a KYC/AML check. If not, we redirect him to the KYC flow. After he has filled in the needed info there, we inform him that our KYC officers are looking at the info and will contact him ASAP with a contract and/or decision. If the user was already known, we simply draw up and send the contract as before.

## The KYC flow in the app
This flow pulls all relevant information from IYO. It uses this information (country of residence, mostly) to consult the information needed to perform a KYC/AML check by our provider Trulioo. Missing information (government ID number, date of birth, province / state / otherthings) will have to be requested from the user. Any fields that are supported by IYO (e.g., date of birth) can get stored in IYO, other information is not stored (to avoid "Are we allowed to store docID numbers?")  but only passed on to the KYC provider. Following the decision by the KYC providers, the user is informed and receives a contract if applicable.

## The KYC flow / results in the backend
The KYC provider returns a result that we pass on to "our KYC officers for review". The KYC officers decide yes/no(/more info needed). If the officers say yes, the TF member is now flagged as being KYC/AML approved (and, depending on if the KYC flow was part of a buying tokens process, the contract is sent to the user). If the officers say no, we flag the user as being barred from buying tokens / receiving a wallet (and inform the user). If we need more info, the KYC officers can contact the user to try and get more info to make a decision. After having reached a decision, the KYC officers delete the information received from Trulioo.
> Yes, this means users can continuously create new ID's in the hope that one of them will finally pass a KYC/AML check. This implies that we might want to perform some logic on our backend DB to check if a user like that already exists. However, in the future, such an operation will no longer be possible as we will no longer be able to access user data. Let's not try to solve this before we can or make an issue where there is none.

That's about it for the most important stuff.

> ## Implications of KYC
> We no longer need to have a user buy a token to receive a wallet. E.g., we _could_ just have a "Get KYC'ed" button that results in a user receiving an active wallet. However, KYC checks cost money, and we have previously stated you need to buy at least one token to obtain a wallet (tokens are currently cheap, meaning the financial cost of one token is less than the amount of effort users have to spend. Should tokens cost 10K USD, the story would be different). Hence, we will keep the KYC flow as part of the "Buy tokens" step until we have a compelling reason to change that.

**Comment by dvorkbod** [comment]
Missing info:

Backend logic regarding the database needs of the different countries in a structured way - maybe Lee already has some of this from his test API calls? It's in the Trulioo documents, in any case.

Insight from my end into how "complicated" flows, such as the one above, are supported by the TF app. After all, a KYC flow is now both callable / pushable to legacy users and to be contained in the "buy tokens" flow. What is a good way to solve this?

**Comment by dvorkbod** [comment]
Related flow on the staging flow designer:
buy_tokens_ITO_v3_with_KYC

**Comment by lucasvanhalst** [comment]
1) ask nationality
2) api call -> get minimum required fields
3) Ask required info (that we do not already know from itsyou.online)
4) Verify this required info
5) parse any errors from 4) and return this to the user, if there are errors, goto 3), else goto 6)
6) Mark this user as "to verify" by our "KYC officers" with the results from 5)
7) KYC officer marks the user as verified / need more info / too high risk (whatever we will call it)

**Comment by dvorkbod** [comment]
Yes, that's the general idea for the KYC (sub)flow!
Do you mean (2) is already an API call? Or that you are preparing the
subsequent API call that passes user ID info?
Verifying required info will probably not be done on our side, as in "is
this a valid ID number according to some sanity check?".
Thanks for the schematic!

**Comment by lucasvanhalst** [comment]
Questions / remarks:

We will probably have to change the 'buy tokens' flow so that it cannot be started unless you have been KYC/AML approved. We will have to remove the name, address, .. (except for shipping address) input fields.  Else they can enter a different name which would defeat the purpose of the whole KYC thing.

This brings us to another minor problem, what about people who want to buy tokens for other people? Do we just tell them to reinstall / install on a different device / order tokens under their name?

**Comment by dvorkbod** [comment]
The name and so on should get pulled from the IYO account. After all, we don't want people filling in false names there anyway.
As for the location of the flow, it seems prudent to make it part of the "buy tokens". But they obviously only have to go through it the first time they sign up for tokens...

**Comment by lucasvanhalst** [comment]
@bart-at-mobicage do we store the request/response from trulioo, or do we just store the transaction id and if we need the info fetch it afterwards via their api?

**Comment by lucasvanhalst** [comment]
New order tokens flow:

- Check KYC status:
  - Approved: show confirmation screen with kyc-approved information, then continue flow
  - Not approved: start kyc flow, show message indicating that once we have KYC approved the user, he can start the order tokens flow again
- Flow done:
  - when KYC approved, generate contract
  - else send message to support to tell them to verify this user

When we mark a user as KYC approved, send a message to the user stating this.

**Comment by dvorkbod** [comment]
It was OK until last week. Following today's call, the process has changed a bit.
We will send out contracts with the conditional clause that it is only valid on passing KYC/AML checks.

People will be contacted in app, asked if they want to have a video call and we will push them the KYC flow to their app. This implies that Sabrina, Gloria, Jenevie will have to be able to push that flow to users on the click of a button, and that we need a system inn place to store the decisions of the KYC/AML checks. The flow should get prepopulated as much as possible (names, address, whatever else we know).

I am updating the flow in the staging version now.
Asking people to type lines from the back of their ID card is not needed if we are going to ask them for pictures of their ID card.

**Comment by dvorkbod** [comment]
The flow to be deployed today is 
"buy_tokens_ITO_v3_KYCmention"
It simply mentions the KYC procedure.

Issues:
1. I have just tested it on staging (page 2 of the app, anchor labeled "Async"), but I did not receive a purchase agreement.
2. The backend should register a new user has to go through KYC, if applicable.
3. If a new user does not initiate a KYC procedure with 3 days, the user should be flagged as someone to be contacted for this reason
4. The same location where orders are being marked as "paid" should have a checkbox / info field that allows payment admins to verify the KYC status of the user.

Tagging @bart-at-mobicage  @lucasvanhalst

**Comment by lucasvanhalst** [comment]
Live on staging

But I've heard we're changing KYC provider, so most of this was for nothing

**Comment by dvorkbod** [comment]
The general framework will remain the same, even when switching provider.

**Comment by lucasvanhalst** [comment]
for onfido, what needs to be validated?  https://documentation.onfido.com/?shell#identity-verification

**Comment by lucasvanhalst** [comment]
This is finally done 🎊
